cmd/auth/auth.go

```diff
@@ -0,0 +1,239 @@
+	if err := json.NewDecoder(res.Body).Decode(&responseData); err != nil {
+		return nil, fmt.Errorf("could not decode JSON response: %s", err.Error())
+	}
+
+	return &responseData, nil
+}
+
+func pkce(length int) (verifier, challenge string, err error) {
+	p := make([]byte, length)
+	if _, err := io.ReadFull(rand.Reader, p); err != nil {
+		return "", "", fmt.Errorf("failed to read random bytes: %s", err.Error())
+	}
+	verifier = base64.RawURLEncoding.EncodeToString(p)
+	b := sha256.Sum256([]byte(verifier))
+	challenge = base64.RawURLEncoding.EncodeToString(b[:])
+	return verifier, challenge, nil
+}
+
+func shutdown(server *http.Server) {
+	go func() {
+		err := server.Shutdown(context.Background())
+		if err != nil {
+			panic(err.Error())
+		}
+	}()
+}
+
+type AccessTokenResponse struct {
+	AccessToken  string     `json:"access_token"`
+	RefreshToken string     `json:"refresh_token"`
+	TokenType    string     `json:"token_type"`
+	ExpiresIn    int        `json:"expires_in"`
+	ExpiresAt    int        `json:"expires_at"`
+	User         UserSchema `json:"user"`
+}
+
+type UserSchema struct {
+	ID                     string                 `json:"id"`
+	Aud                    string                 `json:"aud,omitempty"`
+	Role                   string                 `json:"role"`
+	Email                  string                 `json:"email"`
+	EmailConfirmedAt       string                 `json:"email_confirmed_at"`
+	Phone                  string                 `json:"phone"`
+	PhoneConfirmedAt       string                 `json:"phone_confirmed_at"`
+	ConfirmationSentAt     string                 `json:"confirmation_sent_at"`
+	ConfirmedAt            string                 `json:"confirmed_at"`
+	RecoverySentAt         string                 `json:"recovery_sent_at"`
+	NewEmail               string                 `json:"new_email"`
+	EmailChangeSentAt      string                 `json:"email_change_sent_at"`
+	NewPhone               string                 `json:"new_phone"`
+	PhoneChangeSentAt      string                 `json:"phone_change_sent_at"`
+	ReauthenticationSentAt string                 `json:"reauthentication_sent_at"`
+	LastSignInAt           string                 `json:"last_sign_in_at"`
+	AppMetadata            map[string]interface{} `json:"app_metadata"`
+	UserMetadata           map[string]interface{} `json:"user_metadata"`
+	Factors                []MFAFactorSchema      `json:"factors"`
+	Identities             []interface{}          `json:"identities"`
+	BannedUntil            string                 `json:"banned_until"`
+	CreatedAt              string                 `json:"created_at"`
+	UpdatedAt              string                 `json:"updated_at"`
+	DeletedAt              string                 `json:"deleted_at"`
+}
+
+type MFAFactorSchema struct {
+	ID           string `json:"id"`
+	Status       string `json:"status"`
+	FriendlyName string `json:"friendly_name"`
+	FactorType   string `json:"factor_type"`
+}
```

Comment:
Very nice: thanks for getting these all interpolated out.

Usually structs appear towards the top of a file, or at least before the functions that use them. Feel free to drop these in a different file within the auth package if that makes sense too.

These all also probably make sense as private structs to the auth package: I doubt we'll need these exported by `auth` package consumers:

```go
type accessTokenResponse struct {
```

Reply:
https://github.com/open-sauced/pizza-cli/pull/21/commits/6f02e2c54f6f3700195dd842760b067b5fdc29de